feat: add support for generating embeddings for external documents (#442)
timescale/pgai: docs/vectorizer/api-reference.md

```bash
export AWS_ACCESS_KEY_ID='your_access_key'
export AWS_SECRET_ACCESS_KEY='your_secret_key'
export AWS_REGION='your_region'  # optional
```

Make sure these environment variables are properly set in the environment where the PGAI vectorizer worker runs.

#### Example usage

```sql
SELECT ai.create_vectorizer(
    'my_table'::regclass,
    loading => ai.loading_uri('file_uri_column_name'),
    -- other parameters...
);
```

#### Parameters

`ai.loading_uri` takes the following parameters:

| Name | Type | Default | Required | Description |
|------|------|---------|----------|-------------|
| column_name | text | - | ✔ | The name of the column containing the file path. |

#### Returns

A JSON configuration object that you can use in [ai.create_vectorizer](#create-vectorizers).

## Parsing configuration

You use the parsing configuration functions in `pgai` to define how data is parsed after document loading. This is useful if for non-textual formats such as PDF documents.

The parsing functions are:

- [ai.parsing_auto](#aiparsing_auto): Automatically selects the appropriate parser based on file type.

> - [ai.parsing_none](#aiparsing_none): Converts various formats to Markdown.

Reviewer: I think this is a left over mistake. Below we say that this option `skip the parsing step`

```

2. **Google Cloud Storage**:
```bash
export GOOGLE_APPLICATION_CREDENTIALS='/path/to/credentials.json'
```

3. **Azure Blob Storage**:
```bash
export AZURE_STORAGE_CONNECTION_STRING='your_connection_string'
```

Make sure these environment variables are properly set in the environment where the PGAI vectorizer worker runs.

#### Example usage

```sql
SELECT ai.create_vectorizer(
    'my_table'::regclass,
    loading => ai.loading_uri('file_uri_column_name'),
    -- other parameters...
);
```

#### Parameters

`ai.loading_uri` takes the following parameters:

| Name | Type | Default | Required | Description |
|------|------|---------|----------|-------------|
| column_name | text | - | ✔ | The name of the column containing the file path. |

#### Returns

A JSON configuration object that you can use in [ai.create_vectorizer](#create-vectorizers).

## Parsing configuration

You use the parsing configuration functions in `pgai` to define how data is parsed after document loading. This is useful if for non-textual formats such as PDF documents.

The parsing functions are:

- [ai.parsing_auto](#aiparsing_auto): Automatically selects the appropriate parser based on file type.

> - [ai.parsing_none](#aiparsing_none): Converts various formats to Markdown. See [PyMuPDF](https://pymupdf.readthedocs.io/en/latest/) for supported formats.

Reviewer 2: fixed, thanks

- [ai.parsing_pymupdf](#aiparsing_pymupdf): For cases where no parsing is needed.

### ai.parsing_auto

You use `ai.parsing_auto` to automatically select an appropriate parser based on detected file types.
Documents with unrecognizable formats won't be processed and will generate an error (in the `ai.vectorizer_errors` table.

The parser selection works by examining file extensions and content types:
- For PDF files, images, Office documents (DOCX, XLSX, etc.): Uses docling
- For EPUB and MOBI (e-book formats): Uses pymupdf
- For text formats (TXT, MD, etc.): No parser is used (content is read directly)

#### Example usage

```sql
SELECT ai.create_vectorizer(
    'my_table'::regclass,
    parsing => ai.parsing_auto(),
    -- other parameters...
);
```

#### Parameters

`ai.parsing_auto` takes the following parameters:

| Name | Type | Default | Required | Description |
|------|------|---------|----------|-------------|
| None | - | - | - | - |

#### Returns

A JSON configuration object that you can use in [ai.create_vectorizer](#create-vectorizers).

### ai.parsing_none

> You use `ai.parsing_none` to load the data as-is from the source table.

Reviewer 2: done, thanks

#### Returns

A JSON configuration object that you can use in [ai.create_vectorizer](#create-vectorizers).

### ai.loading_document

You use `ai.loading_document` to load the data to embed from a file that is referenced in a column of the source table.
This file path is internally passed to [smart_open](https://github.com/piskvorky/smart_open), so it supports any protocol that smart_open supports. E.g. S3, GCS, etc.
You just need to ensure the vectorizer worker has the correct credentials to access the file.
For S3, you can set the `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` environment variables.

#### Example usage

```sql
SELECT ai.create_vectorizer(
    'my_table'::regclass,
    loading => ai.loading_document('file_uri_column_name'),
    -- other parameters...
);
```

#### Parameters

`ai.loading_document` takes the following parameters:

| Name | Type | Default | Required | Description |
|------|------|---------|----------|-------------|
| column_name | text | - | ✔ | The name of the column containing the file path. |

#### Returns

A JSON configuration object that you can use in [ai.create_vectorizer](#create-vectorizers).

## Parsing configuration

> You use the parsing configuration functions in `pgai` to define the way data, after being loaded, is parsed. This is useful if you want to load data that is not immediately text and therefore needs to be parsed first. I.e. a PDF document.

Reviewer 3: ```suggestion
You use the parsing configuration functions in `pgai` to define how data is parsed after document loading. This is useful if for non-textual formats such as PDF documents.
```

The parsing functions are:

- [ai.parsing_auto](#aiparsing_auto)
- [ai.parsing_none](#aiparsing_none)
- [ai.parsing_docling](#aiparsing_docling)
- [ai.parsing_pymupdf](#aiparsing_pymupdf)

> ### ai.parsing_auto

Reviewer 3: Should we document how the underlying parser is chosen?

Reviewer 4: done ✅

> You use `ai.parsing_auto` to automatically parse the data with a fitting available parser.
> This special parser will decide on the fly which parser fits with each of the documents by guessing the file type. 
> If the type can not be guessed, the document will not be processed and an error will be appended to the vectorizer errors.

Reviewer 3: ```suggestion
You use `ai.parsing_auto` to automatically select an appropriate parser based on detected file types.
Documents with unrecognizable formats won't be processed and will generate an error (in the `ai.vectorizer_errors` table.
```

#### Example usage

```sql
SELECT ai.create_vectorizer(
    'my_table'::regclass,
    parsing => ai.parsing_auto(),
    -- other parameters...
);
```

#### Parameters

`ai.parsing_auto` takes the following parameters:

| Name | Type | Default | Required | Description |
|------|------|---------|----------|-------------|
| None | - | - | - | - |

#### Returns

A JSON configuration object that you can use in [ai.create_vectorizer](#create-vectorizers).

### ai.parsing_none

> You use `ai.parsing_none` to load the data as is from the source table.

Reviewer 3: ```suggestion
You use `ai.parsing_none` to load the data as-is from the source table.
```

> [!NOTE]  

> > This is the default parser; adding this parser or skipping the `parsing` parameter does the same effect.

Reviewer 3: This statement is incorrect? The default parser is `parsing_auto`.

#### Example usage

```sql
SELECT ai.create_vectorizer(
    'my_table'::regclass,
    parsing => ai.parsing_none(),
    -- other parameters...
);
```   

#### Parameters

`ai.parsing_none` takes the following parameters:

| Name | Type | Default | Required | Description |
|------|------|---------|----------|-------------|
| None | - | - | - | - |

#### Returns

A JSON configuration object that you can use in [ai.create_vectorizer](#create-vectorizers).

> ### ai.parsing_docling

Reviewer 3: Should we provide a suggestion on when one should use docling vs pymypdf?

Reviewer 4: done ✅
